src/yb/util/wait_state.cc

```diff
@@ -49,8 +51,40 @@ simple_spinlock* WaitStateInfo::get_mutex() {
   return &mutex_;
 };
 
+void WaitStateInfo::freeze() {
+  freeze_ = true;
+}
+
+void WaitStateInfo::unfreeze() {
+  freeze_ = false;
+}
+
+WaitStateCode WaitStateInfo::get_frozen_state() const {
+  if (frozen_state_code_ != WaitStateCode::Unused) {
+    return frozen_state_code_;
+  }
+  return code_;
+}
+
 void WaitStateInfo::set_state(WaitStateCode c) {
   VLOG(3) << this << " " << ToString() << " setting state to " << util::ToString(c);
+  if (freeze_) {
+    // If this is the first time we are calling set_state after freeze() was called,
+    // we will save the current state before updating it.
+    // If state is not frozen, then we reset frozen_state to Unused.
+    //
+    // This mechanism assumes that when unfreeze is called, the WaitState will be updated at least
+    // once, before the next freeze is called. This should generally be fine as we plan to freeze
+    // when AUHmetadata is pulled -- once every second, while the update should happen more often.
+    // But just something to note. If this is violated, then the 2nd, frozen state will represent
+    // the start of the 1st freeze window, instead of reflecting the start state of the 2nd freeze
+    // window.
```

Comment:
unfreeze() is a static function.  It would be too costly and not feasible to find all instances of WaitStateInfos and reset `frozen_state_code` to 0. 

One option could be to update `frozen_state_code` when we do ToPB(...) as part of the Auh call. But again, that is not guaranteed to run for all WaitStateInfo objects. So, we it doesn't really get rid of this limitation.

IMO, the limitation is ok. As this is generally (at least as of now) a debugging flag to see what effect, if at all, the jagged edge has over the wait-states collected.